internal/net/tcp/dialer.go

```diff
@@ -118,70 +115,88 @@ func NewDialer(opts ...DialerOption) (der Dialer, err error) {
 	return d.dialer
 }
 
-func (d *dialer) lookup(ctx context.Context,
-	addr string) (ips map[int]string, err error) {
-	cache, ok := d.cache.Get(addr)
+func (d *dialer) lookup(ctx context.Context, host string) (*dialerCache, error) {
+	cache, ok := d.cache.Get(host)
 	if ok {
-		return cache.(map[int]string), nil
+		return cache.(*dialerCache), nil
 	}
 
-	r, err := d.der.Resolver.LookupIPAddr(ctx, addr)
+	r, err := d.der.Resolver.LookupIPAddr(ctx, host)
 	if err != nil {
 		return nil, err
 	}
 
-	ips = make(map[int]string, len(r))
-	for i, ip := range r {
-		ips[i] = ip.String()
+	dc := &dialerCache{
+		ips: make([]string, 0, len(r)),
+	}
+	for _, ip := range r {
+		dc.ips = append(dc.ips, ip.String())
 	}
 
-	d.cache.Set(addr, ips)
-	return ips, nil
+	d.cache.Set(host, dc)
+	return dc, nil
 }
 
+// StartDialerCache starts the dialer cache to expire the cache automatically
 func (d *dialer) StartDialerCache(ctx context.Context) {
 	if d.dnsCache && d.cache != nil {
 		d.cache.Start(ctx)
 	}
 }
 
+// DialContext returns the connection or error base on the input.
+// If the DNS cache is enabled, it will lookup the DNS cache in round robin order and return a connection of it.
+// Also if TLS is enabled, it will create a TLS connection for it.
 func (d *dialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
 	return d.GetDialer()(ctx, network, address)
 }
 
-func (d *dialer) cachedDialer(dctx context.Context, network, addr string) (
-	conn net.Conn, err error) {
-	sep := strings.LastIndex(addr, ":")
-
-	if sep < 0 {
-		sep = len(addr)
+func (d *dialer) cachedDialer(dctx context.Context, network, addr string) (conn net.Conn, err error) {
+	host, port, isIP, err := net.Parse(addr)
+	if err != nil {
+		return nil, err
 	}
 
-	ips, err := d.lookup(dctx, addr[:sep])
-	if err == nil {
-		for _, ip := range ips {
-			conn, err = d.der.DialContext(dctx, network, ip+addr[sep:])
-			if err == nil {
-				if d.tlsConfig != nil {
-					return tls.Client(conn, d.tlsConfig), nil
+	if d.dnsCache && !isIP {
+		if dc, err := d.lookup(dctx, host); err == nil {
+			for i := uint32(0); i < dc.Len(); i++ {
+				if conn, err := d.dial(dctx, network, fmt.Sprintf("%s:%d", dc.GetIP(), port)); err == nil {
+					return conn, nil
 				}
-				return conn, nil
 			}
+			d.cache.Delete(host)
+		}
+	}
+	return d.dial(dctx, network, addr)
+}
+
+func (d *dialer) dial(ctx context.Context, network, addr string) (net.Conn, error) {
+	conn, err := d.der.DialContext(ctx, network, addr)
+	if err != nil {
+		defer func(conn net.Conn) {
 			if conn != nil {
 				conn.Close()
 			}
-		}
-		d.cache.Delete(addr[:sep])
+		}(conn)
+		return nil, err
 	}
 
-	conn, err = d.der.DialContext(dctx, network, addr)
 	if d.tlsConfig != nil {
 		return tls.Client(conn, d.tlsConfig), nil
 	}
-	return
+	return conn, nil
+}
+
+func (d *dialer) cacheExpireHook(ctx context.Context, addr string) {
+	if err := safety.RecoverFunc(func() error {
+		_, err1 := d.lookup(ctx, addr)
```

Comment:
please do not use numbering variable
```suggestion
		_, err1 := d.lookup(ctx, addr)
```

Reply:
fixed